### ISSUE TITLE
howto set variables with commands?

### ISSUE THREAD

**Issue by niksfirefly** [DannyBen/opcode#21]
I try to setup
debug: BASH_ENV=debug.sh ./app
and it not working
howto set variables with commands?

regards

I use bashly,opcode,rush keep up good work!

**Comment by DannyBen** [comment]
Hi - glad you find these tools useful. I am also using them all :)
Check out [Fuzzy CD](https://github.com/DannyBen/fuzzycd) - my latest addition to the toolbelt.

All commands in opcode are running with `eval`, so I think this is probably the best way:

```bash
debug: BASH_ENV=debug && echo BASH_ENV is $BASH_ENV
```

**Comment by niksfirefly** [comment]
If you add command at the end
`debug: BASH_ENV=debug && echo BASH_ENV is $BASH_ENV && ./app`
app cant see $BASH_ENV

**Comment by DannyBen** [comment]
Well, two things - first a solution, then a suggestion:

Think of an opcode line as a mini bash script. If you put the same commands in a bash script they will work exactly the same:

```bash
# runme.sh
#!/usr/bin/env bash
BASH_ENV=debug
./test.sh

# test.sh
#!/usr/bin/env bash
echo BASH_ENV is $BASH_ENV
```

```bash
# running it does not pass on the variable.
$ ./runme.sh
BASH_ENV is
```

This is because of the subshell that the script you are calling is using.

To fix it, simply use `export`, like you would in a bash script.

```bash
# op.conf
debug: export BASH_ENV=debug && ./test.sh
```

Now for the suggestion.

If the `./app` you are running is something you developed, I suggest you consider adding an argument that is equivalent to the environment variable, so you can run something like:

```
$ ./app --bash-env debug
```

**Comment by niksfirefly** [comment]
Well that dont work with opcode
`debug: export BASH_ENV=debug && ./test.sh`

BASH_ENV is a special var which is used by Bash to define an init file to read before executing a script
https://www.shell-tips.com/bash/debug-script/#step-1-use-a-consistent-debug-library
I want to separate debugging from my app to make them smaller.

Of  course I can do 
export BASH_ENV=debug && ./test.sh
but I try to create development template for all my apps with opcode, bashly and cookiecutter to make all simple and easy to remember no matter what script I am creating

**Comment by DannyBen** [comment]
Then the question is not about *any* environment variables, but this particular bash behavior.
In this case, I don't know what to tell you as I am not using this functionality, it seems like a good way to complicate things, and I never needed this.

Opcode just uses `eval` to run your line, as can be seen here:

https://github.com/DannyBen/opcode/blob/20c14188f1da213e4b711edf50afdecd89be94b1/op#L86-L90

If the task you are trying to complete works with `eval`, it will work with `opcode`, otherwise, it wont.

**Comment by DannyBen** [comment]
But that said - from my tests, it works just the same with `BASH_ENV`:

```bash
# op.conf
debug: export BASH_ENV=init.sh && ./test.sh

# init.sh
#!/usr/bin/env bash
echo "init here"

# test.sh
#!/usr/bin/env bash
echo BASH_ENV is $BASH_ENV
``

and running the opcode:

```
$ op debug
init here
BASH_ENV is init.sh
```

**Comment by niksfirefly** [comment]
Strange thing happen
I moved that command to the top of op.conf file
`deb: export BASH_ENV=debug && ./test.sh`
and now it works

Then I found out that that command deb is not working when is below debug
```
debug: export BASH_ENV=debug && ./test.sh
deb: export BASH_ENV=debug && ./test.sh
```
but 
```
deb: export BASH_ENV=debug && ./test.sh
debug: export BASH_ENV=debug && ./test.sh
```
deb command works when is above debug
strange!

**Comment by DannyBen** [comment]
Yeah, not strange but an overzealous feature.
Opcode supports partial match, so it gets the first command that matches.
I need to see if I can easily change it so it favors exact match before any partial. That is a different issue. (I will open).

Can we close this issue then?

**Comment by niksfirefly** [comment]
Thx

**Comment by DannyBen** [comment]
I just released version 0.6.1 - it will now work (`op deb`) regardless of the order in the file.
Exact match commands will be favored over partial match.
